Docs: Update Loki documentation (#31867)
grafana/grafana: docs/sources/datasources/loki.md

weight = 800
+++

# Loki data source

> Grafana ships with built-in support for Loki, an open source log aggregation system by Grafana Labs. Add it as a data source and you are ready to build dashboards or query your log data in [Explore]({{< relref "../explore" >}}).

Reviewer: ```suggestion
Grafana ships with built-in support for Loki, an open source log aggregation system by Grafana Labs. This topic explains options, variables, querying, and other options specific to this data source.

Add it as a data source and you are ready to build dashboards or query your log data in [Explore]({{< relref "../explore" >}}). Refer to [Add a data source]({{< relref "add-a-data-source.md" >}}) for instructions on how to add a data source to Grafana. Only users with the organization admin role can add data sources.
```

> This topic explains options, variables, querying, and other options specific to this data source. Refer to [Add a data source]({{< relref "add-a-data-source.md" >}}) for instructions on how to add a data source to Grafana. Only users with the organization admin role can add data sources.

Reviewer: Move content up, as shown.

> > **Note:** If you have issues with getting this data source to work and Grafana is giving you undescriptive errors, then check your log file (try looking in /var/log/grafana/grafana.log).

Reviewer: ```suggestion
> **Note:** To troubleshoot configuration and other issues, check the log file (/var/log/grafana/grafana.log).
```

Reviewer 2: This path is not going to work for all users. On my Windows machine, it's in grafana/data/log/grafana.log. I don't have a var folder.

## Loki settings

To access Loki settings, click the **Configuration** (gear) icon, then click **Data Sources**, and then click the Loki data source.

| Name                  | Description                                                                                                                                   |
| --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
| `Name`                | The data source name. This is how you refer to the data source in panels, queries, and Explore.                                               |
| `Default`             | Default data source means that it will be pre-selected for new panels.                                                                        |
| `URL`                 | The URL of the Loki instance, e.g., `http://localhost:3100`                                                                                   |
| `Whitelisted Cookies` | Grafana Proxy deletes forwarded cookies by default. Specify cookies by name that should be forwarded to the data source.                      |

> | `Maximum lines`       | Upper limit for number of log lines returned by Loki (default is 1000). Decrease if your browser is sluggish when displaying logs in Explore. |

Reviewer: ```suggestion
| `Maximum lines`       | Upper limit for the number of log lines returned by Loki (default is 1000). Lower this limit if your browser is sluggish when displaying logs in Explore. |
```

| `Query type`       | Choose the type of query to run. The instant type queries against a single point in time. We are using "To" time from the time range. The range type queries over the selected range of time. |
| `Line limit`       | Upper limit for number of log lines returned by query. The default is the Maximum lines limit set in Loki settings.                                                                                         |
| `Legend`           | Available only in Dashboard. Controls the name of the time series, using name or pattern. For example `{{hostname}}` is replaced with the label value for the label `hostname`.                             |

### Log browser

With Loki log browser you can easily navigate trough your list of labels and values and construct the query of your choice. Log browser has multi-step selection:
1. Choose the labels you would like to consider for your search.
2. Pick the values for selected labels. Log browser supports facetting and therefore it shows you only possible label combinations. 

> 3. Choose the type of query - logs query or rate metrics query. Additionaly, you can also validate selector.

Author: ```suggestion
3. Choose the type of query - logs query or rate metrics query. Additionally, you can also validate selector.
```

The new field with the link shown in log details:
{{< docs-imagebox img="/img/docs/v65/loki_derived_fields_detail.png" class="docs-image--no-shadow" caption="Screenshot of the derived field in log detail" >}}

## Loki Query Editor

> Loki query editor can be used to create both log and metric queries.

Reviewer: ```suggestion
Use the Loki query editor to create both log and metric queries.
```

| Name               | Description                                                                                                                                                                                                 |
| ------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `Query expression` | Loki query expression, check out the [LogQL documentation](https://grafana.com/docs/loki/latest/logql/).                                                                                                    |

Reviewer: ```suggestion
| `Query expression` | Loki query expression, refer to the [LogQL documentation](https://grafana.com/docs/loki/latest/logql/) for more information.                                                                                                    |
```

> | `Query type`       | Choose the type of query you would like to run. An instant query queries against a single point in time. We are using "To" time from the time range. A range query queries over the selected range of time. |

Reviewer: ```suggestion
| `Query type`       | Choose the type of query to run. The instant type queries against a single point in time. We are using "To" time from the time range. The range type queries over the selected range of time. |
```

| `Line limit`       | Upper limit for number of log lines returned by query. The default is the Maximum lines limit set in Loki settings.                                                                                         |
| `Legend`           | Available only in Dashboard. Controls the name of the time series, using name or pattern. For example `{{hostname}}` is replaced with the label value for the label `hostname`.                             |

### Log browser

// TODO David

## Querying with Loki

There are two types of LogQL queries:

> - **Log queries** that return the contents of log lines.

Reviewer: ```suggestion
- **Log queries:** returns the contents of log lines.
```

Reviewer: Log queries do not need to be bold.

Author: Removed bold. Plus this was changed to `Log queries -`

> - **Metric queries** that extend log queries and calculate sample values based on the content of logs from a log query.

Reviewer: ```suggestion
- **Metric queries:** extends log queries and calculate sample values based on the content of logs from a log query.
```

> - **Metric queries** that extend log queries and calculate sample values based on the content of logs from a log query.

Reviewer: Metric queries do not need to be bold.

### Log Queries

Querying and displaying log data from Loki is available via [Explore]({{< relref "../explore" >}}), and with the [logs panel]({{< relref "../panels/visualizations/logs-panel.md" >}}) in dashboards. Select the Loki data source, and then enter a [LogQL](https://grafana.com/docs/loki/latest/logql/#log-queries) query to display your logs.

> A log query consists of two parts: **log stream selector**, and a **log pipeline**. For performance reasons you need to start by choosing a log stream by selecting a log label.

Reviewer 2: Why are the terms bold? If they are not on the UI, they should not be bold. Apply italic if you want to emphasize them.

Author: Removed bold. I guess, they don't need to be bold.

> A log query consists of two parts: **log stream selector**, and a **log pipeline**. For performance reasons you need to start by choosing a log stream by selecting a log label.

Reviewer: ```suggestion
A log query consists of two parts: **log stream selector**, and a **log pipeline**. For performance reasons begin by choosing a log stream by selecting a log label.
```

> ### Log Context

Reviewer 2: ```suggestion
### Log context
```

> When using a search expression as detailed above, you now have the ability to retrieve the context surrounding your filtered results.

Reviewer 2: ```suggestion
When using a search expression as detailed above, you can retrieve the context surrounding your filtered results.
```

By clicking the `Show Context` link on the filtered rows, you'll be able to investigate the log messages that came before and after the
log message you're interested in.

> > **Note:** This feature is only available in Grafana v6.3+

Reviewer: Can we remove this note? 6.3 is really old.

Author: removed
